presto-raptorx/src/main/java/com/facebook/presto/raptorx/storage/OrcPageSource.java

```diff
@@ -0,0 +1,262 @@
+            }
+            else if (this.columnIndexes[i] == BUCKET_NUMBER_COLUMN) {
+                constantBlocks[i] = buildSingleValueBlock(columnTypes.get(i), (long) bucketNumber);
+            }
+        }
+
+        this.systemMemoryContext = requireNonNull(systemMemoryContext, "systemMemoryContext is null");
+    }
+    @Override
+    public long getCompletedBytes()
+    {
+        return orcDataSource.getReadBytes();
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return orcDataSource.getReadTimeNanos();
+    }
+
+    @Override
+    public long getSystemMemoryUsage()
+    {
+        return systemMemoryContext.getBytes();
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return closed;
+    }
+
+    @Override
+    public Page getNextPage()
+    {
+        try {
+            batchId++;
+            int batchSize = recordReader.nextBatch();
+            if (batchSize <= 0) {
+                close();
+                return null;
+            }
+            long filePosition = recordReader.getFilePosition();
+
+            Block[] blocks = new Block[columnIndexes.length];
+            for (int fieldId = 0; fieldId < blocks.length; fieldId++) {
+                Type type = types.get(fieldId);
+                if (constantBlocks[fieldId] != null) {
+                    blocks[fieldId] = constantBlocks[fieldId].getRegion(0, batchSize);
+                }
+                else if (columnIndexes[fieldId] == ROW_ID_COLUMN) {
+                    blocks[fieldId] = buildSequenceBlock(filePosition, batchSize);
+                }
+                else {
+                    blocks[fieldId] = new LazyBlock(batchSize, new OrcBlockLoader(columnIndexes[fieldId], type));
+                }
+            }
+
+            return new Page(batchSize, blocks);
+        }
+        catch (IOException | RuntimeException e) {
+            closeWithSuppression(e, this);
+            throw new PrestoException(RAPTOR_STORAGE_ERROR, e);
+        }
+    }
+
+    @Override
+    public void close()
+    {
+        closed = true;
+
+        try {
+            recordReader.close();
+        }
+        catch (IOException e) {
+            throw new PrestoException(RAPTOR_STORAGE_ERROR, e);
+        }
+    }
+
+    @Override
+    public String toString()
+    {
+        return toStringHelper(this)
+                .add("columnNames", columnIds)
+                .add("types", types)
+                .toString();
+    }
+
+    @Override
+    public void deleteRows(Block rowIds)
+    {
+        for (int i = 0; i < rowIds.getPositionCount(); i++) {
+            long rowId = BIGINT.getLong(rowIds, i);
+            rowsToDelete.set(toIntExact(rowId));
+        }
+    }
+
+    @Override
+    public CompletableFuture<Collection<Slice>> finish()
+    {
+        checkState(chunkRewriter.isPresent(), "chunkRewriter is missing");
+        return chunkRewriter.get().rewrite(rowsToDelete);
```

Comment:
`rewrite` -> `rewriteIfNecessary` or
return directly if `rowsToDelete` is empty and enforce `rowsToDelete` is non-empty inside `rewrite`